src/notebook/codemirror/cells/editor.ts

```diff
@@ -250,14 +250,17 @@ class CodeMirrorCellEditorWidget extends CodeMirrorWidget implements ICellEditor
       return this.onTabEvent(event, ch, line);
     }
 
-    if (line === 0 && ch === 0 && event.keyCode === UP_ARROW) {
+    // If the shift key is pressed, don't move to top cell
+    if (line === 0 && ch === 0 && event.keyCode === UP_ARROW && event.shiftKey === false) {
       this.edgeRequested.emit('top');
```

Comment:
Since you want to `return` either way, I would suggest instead, doing this (obviously for both `top` and `bottom`):

```
if (!event.shiftKey) {
  this.edgeRequested.emit('top');
}
return;
```